src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1105,6 +1105,14 @@ private void ImportCall(ILOpcode opcode, int token)
         private LLVMValueRef LLVMFunctionForMethod(MethodDesc callee, StackEntry thisPointer, bool isCallVirt)
         {
             string calleeName = _compilation.NameMangler.GetMangledMethodName(callee).ToString();
+
+            // Sealed methods must not be called virtually due to sealed vTables, so call them directly
+            if(callee.IsFinal || callee.OwningType.IsSealed())
```

Comment:
Note that the "`IsVirtual` and opcode is callvirt -> this is a virtual call" pattern is duplicated in this codebase:

https://github.com/dotnet/corert/blob/84abd2b56c32550cc7a2bbbb66968752a62b4992/src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs#L1363-L1366

We might have to also update that. But it's also possible it's just another piece of useless code that should be deleted. There's plenty of code [like that](https://github.com/dotnet/corert/blob/84abd2b56c32550cc7a2bbbb66968752a62b4992/src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs#L1085-L1089) in this file so it's starting to be hard to tell.

Reply:
I'm not quite sure I understand what you mean about updating it... do you just want something that captures that it's really virtual in one of those cases to pass it on to the other?

Reply:
I meant two things:

* by "the pattern is duplicated" I meant that there is still code that is going to report method calls as virtual, even though they got devirtualized after your change. This wastes space in the (sealed) vtable since the slot is not actually used (because codegen made a direct call). It's not clear why we need to do the `AddVirtualMethodReference` there, since this seems to be covered by other calls to `AddVirtualMethodReference`.
* by "it's also possible it's just another piece of useless code that should be deleted" I meant that maybe after #4999 all virtual method usage is tracked through references to `WebAssemblyVTableSlotNode` and `AddVirtualMethodReference` can be deleted without replacement.